### DISCUSSION TITLE
`str expand`: equivalent of `cat ~/{this,and,that}`

### DISCUSSION THREAD

**Discussion by atahabaki** [nushell/nushell#9277] [Ideas]
I'm considering creating a command or extension command for Nushell. Currently, I'm using `str expand` in my code, but I'm unsure if it's the right choice. I believe that `glob` is more appropriate for cases where files actually exist, like git commits or editing files. For cases where files don't exist, it doesn't seem logical to include them in `glob`. I would appreciate your thoughts on this. I'm working on it and planning to make a pull request.

Here is the example usage of `str expand`:

```nushell
^mkdir -p ("~/.config/{nvim,waybar}" | str expand)
```

The parameter doesn't necessarily have to be a path; it could be anything. In general, it can take a string in the form of `{this,and,that}` or a similar format. So, instead of restricting it to file paths, it can be used in a broader sense to accommodate various types of strings. I hope this clarifies the flexibility of the parameter.

Please let me know what do you think.

**Comment by amtoine** [comment]
@atahabaki 

you mean `path expand`, right? :open_mouth:

**Reply by fdncred** [reply]
As I was reading the post, I was thinking the exact same thing. It seems like it's a better fit for `path expand`, which we already have. I'm thinking that this functionality could be added to it without any new parameters.

**Reply by atahabaki** [reply]
Whole point of `str expand` is making below code much less repetitive,

```
^mkdir -p ~/.config/nvim ~/.config/waybar
```

like bash does:

```bash
mkdir -p ~/.config/{nvim,waybar}
# above code is =
# mkdir -p ~/.config/nvim ~/.config/waybar
```

If `path expand` does this, could you give me an example?

```nu
"~/.config/{nvim,waybar}" | path expand
# /home/atahabaki/.config/{nvim,waybar}
```

Could you please provide me with an example to demonstrate how the "path expand" command works? I'm interested in learning more about its friendly functionality.

**Reply by fdncred** [reply]
You can see the examples with `help path expand` or `path expand --help`. I'm not saying it does this now, I'm saying it's a better fit than `str expand`.

**Reply by amtoine** [reply]
i'm confused... is `str expand` even an existing command? :confused:

**Reply by fdncred** [reply]
@amtoine I think @atahabaki is proposing a new command named `str expand`. I'm voting against that because I think `path expand` is a better fit for the functionality.

**Reply by atahabaki** [reply]
```nu
^mkdir (["./a1","./a2"] | each {|p| path expand})
# Output as expected:
# a1 and a2 dirs created.
```

```nu
^mkdir (["./a1{b1,b2,b3}"] | each {|p| path expand})
# Output:
# `a1{b1,b2,b3}`, just one dir created. What this command should create 3 dirs.
```

This is why I need another command. Takes a string, generates a list of string according to the given pattern/string.

**Reply by atahabaki** [reply]
> i'm confused... is `str expand` even an existing command? :confused: 

no, it is not. I'm working on a command that does this:

**Reply by amtoine** [reply]
oooooooh okey, my bad, now it all makes sense :relieved:

**Comment by atahabaki** [comment]
I'd like to share some examples of a command I'm currently working on, which I'll refer to as `str expand` for now.

#### 1. Basic Brace Expansion

```nu
echo ("{apple,banana,cherry}" | str expand)
# Output: apple banana cherry
```

In this example, the brace expansion pattern `{apple,banana,cherry}` is expanded by `str expand`, resulting in the output `apple banana cherry`.

Another example:

```nu
echo ("~/Desktop/{file1,file2,file3}.txt" | str expand)
# Output: ~/Desktop/file1.txt ~/Desktop/file2.txt ~/Desktop/file3.txt
```

In this example, the brace expansion pattern `~/Desktop/{file1,file2,file3}.txt` represents multiple file paths with a common directory. `str expand` expands the pattern, resulting in the output of the expanded file paths.

#### 2. Numeric Range Expansion:

```nu
echo ("{1..5}" | str expand)
# Output: 1 2 3 4 5
```

In this example, the brace expansion pattern {1..5} represents a numeric range from 1 to 5. `str expand` expands the pattern, resulting in the output `1 2 3 4 5`.

> Note: I'm still considering whether this feature should be included directly within the command.

#### 3. Combined Brace Expansion:

```nu
echo ("{A{1,2},B{3,4}}" | str expand)
# Output: A1 A2 B3 B4
```

Here, the brace expansion pattern `{A{1,2},B{3,4}}` contains nested brace expansions. `str expand` expands the pattern, resulting in the output `A1 A2 B3 B4`.

These examples demonstrate the various use cases of `str expand`, including basic brace expansion, numeric range expansion, and combined brace expansion.

I'm excited to share that I'm currently working on the `str expand` command, but I could use some guidance regarding its naming and placement. I would greatly appreciate your input and feedback on where you think this command should be located. Additionally, any thoughts or suggestions you have about the command itself are most welcome. Your valuable insights will help shape the development of this feature. Thank you in advance for taking the time to share your feedback, and I look forward to hearing your thoughts!

**Reply by fdncred** [reply]
Thanks for providing your examples. Those should all go in the examples function of the command itself because it explains a lot. I think I could live with `str expand` after seeing these examples.

If it's a `str` subcommand, it should go where the other `str` subcommands are. I'm anxious to see a PR now. Thanks!

**Reply by atahabaki** [reply]
Here it is, #9290. I can't wait to see it in the `nushell v0.80.1`. For now, I have not implemented the example 2 (Numeric Range Expansion), but I'll.

**Comment by tgross35** [comment]
Is there any easy way to combine `str expand` with glob expansion? E.g. the following expansion happens:

```nu
> echo ("somepath/{Cargo.*,src}" | str expand)
 0   somepath/Cargo.*
 1   somepath/src
```

But ideally it returns `somepath/Cargo.toml`, `somepath/Cargo.lock`, and `somepath/src`.

**Reply by atahabaki** [reply]
By iterating over the list and then using flatten...

For example,

```
"somepath/{Cargo.*,src}" | str expand | each {glob $in} | flatten
```
Result:

```
╭───┬────────────────────────────────────────────────╮
│ 0 │ /data/data/com.termux/files/home/xd/Cargo.lock │
│ 1 │ /data/data/com.termux/files/home/xd/Cargo.toml │
│ 2 │ /data/data/com.termux/files/home/xd/src        │
╰───┴────────────────────────────────────────────────╯
```
